src/goto-instrument/reachability_slicer.cpp

```diff
@@ -20,41 +24,62 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "full_slicer_class.h"
 #include "reachability_slicer_class.h"
 
-void reachability_slicert::fixedpoint_assertions(
+/// Get the set of nodes that correspond to the given criterion, or that can
+/// appear in concurrent execution. None of these should be sliced away so
+/// they are used as a basis for the search.
+/// \param is_threaded Instructions that might be executed concurrently
+/// \param criterion The criterion we care about
+std::vector<reachability_slicert::cfgt::node_indext>
+reachability_slicert::get_sources(
   const is_threadedt &is_threaded,
   slicing_criteriont &criterion)
 {
-  queuet queue;
+  std::vector<cfgt::node_indext> sources;
+  for(const auto &e_it : cfg.entry_map)
+  {
+    if(criterion(e_it.first) || is_threaded(e_it.first))
+      sources.push_back(e_it.second);
+  }
 
-  for(cfgt::entry_mapt::iterator
-      e_it=cfg.entry_map.begin();
-      e_it!=cfg.entry_map.end();
-      e_it++)
-    if(criterion(e_it->first) ||
-       is_threaded(e_it->first))
-      queue.push(e_it->second);
+  return sources;
+}
 
-  while(!queue.empty())
-  {
-    cfgt::entryt e=queue.top();
-    cfgt::nodet &node=cfg[e];
-    queue.pop();
+/// Perform backwards depth-first search of the control-flow graph of the
+/// goto program, starting from the nodes corresponding to the criterion and
+/// the instructions that might be executed concurrently. Set reaches_assertion
+/// to true for every instruction visited.
+/// \param is_threaded Instructions that might be executed concurrently
+/// \param criterion the criterion we are trying to hit
+void reachability_slicert::fixedpoint_to_assertions(
+  const is_threadedt &is_threaded,
+  slicing_criteriont &criterion)
+{
+  std::vector<cfgt::node_indext> src = get_sources(is_threaded, criterion);
 
-    if(node.reaches_assertion)
-      continue;
+  std::vector<cfgt::node_indext> reachable = cfg.get_reachable(src, false);
+  for(const auto index : reachable)
+    cfg[index].reaches_assertion = true;
+}
 
-    node.reaches_assertion=true;
+/// Perform backwards depth-first search of the control-flow graph of the
```

Comment:
The comment for this and the above method are identical, I guess this should say "perform *forwards* depth first..."?